From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ian=20Casta=C3=B1o?= <ianfe@martina.com.co>
Date: Tue, 14 Jun 2022 21:16:46 -0500
Subject: [PATCH] Replaced unsupported bungeecord chat version


diff --git a/build.gradle.kts b/build.gradle.kts
index 1c1b604198e13b8e955e5413a7ed595370b67329..44c6033b851d757bd7d6641cc68d2398bef68a3e 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -25,7 +25,7 @@ dependencies {
     // api dependencies are listed transitively to API consumers
     api("com.google.guava:guava:31.0.1-jre")
     api("com.google.code.gson:gson:2.8.9")
-    api("net.md-5:bungeecord-chat:1.16-R0.4-deprecated+build.6") // Paper
+    api("net.md-5:bungeecord-chat:1.16-R0.5-SNAPSHOT") // Paper
     api("org.yaml:snakeyaml:1.33")
     // Paper start
     api("com.googlecode.json-simple:json-simple:1.1.1") {
