From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TomateDeveloper <ianfe@martina.com.co>
Date: Fri, 11 Nov 2022 19:40:24 -0500
Subject: [PATCH] Added find player by ID handle


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index fc0b33272114c13dff6f5a5795e190fda79c071c..ef4feccc746891e490ca881d6b84b1e6ab64b09d 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -556,6 +556,15 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @Nullable
     public Player getPlayerExact(@NotNull String name);
 
+    /**
+     * Gets the player with a previous registered custom id.
+     *
+     * @param id to retrieve
+     * @return a player object if one was found, null otherwise
+     */
+    @Nullable
+    public Player getPlayerByIdentifier(String id); // Sinfonia - Get by player id
+
     /**
      * Attempts to match any players with the given name, and returns a list
      * of all possibly matches.
