From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TomateDeveloper <ianfe@martina.com.co>
Date: Fri, 11 Nov 2022 19:40:24 -0500
Subject: [PATCH] Added find player by ID handle


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 6549c3edea507ce235e855ae534e7778bb5ca999..425f68efce09359d976c3e3d178cbac456d6cc82 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -556,6 +556,15 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @Nullable
     public Player getPlayerExact(@NotNull String name);
 
+    /**
+     * Gets the player with a previous registered custom id.
+     *
+     * @param id to retrieve
+     * @return a player object if one was found, null otherwise
+     */
+    @Nullable
+    public Player getPlayerByIdentifier(String id); // Sinfonia - Get by player id
+
     /**
      * Attempts to match any players with the given name, and returns a list
      * of all possibly matches.
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 9c2dcca0b6e34a3083226360575f0956a077d767..ef6f1a7bc47382b8c14d373efd355d9655d6012d 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2921,6 +2921,20 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     public boolean hasFlyingFallDamage();
 
+    /**
+     +     * Allows to obtain player custom identifier.
+     +     *
+     +     * @return identifier retrieve
+     +     */
+    public String getPlayerIdentifier(); // Sinfonia - Get player by identifier
+
+    /**
+     * Allows to set player custom identifier.
+     *
+     * @param identifier to set
+     */
+    public void setDatabaseIdentifier(String identifier); // Sinfonia - Get player by identifier
+
     /**
      * Creates debug block highlight on specified block location and show it to this player.
      * <p>
