From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ian=20Casta=C3=B1o?= <ianfe@martina.com.co>
Date: Thu, 23 Jun 2022 12:15:23 -0500
Subject: [PATCH] Fixed bukkit testing


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index e5943f1313f97dd6a353873d54424ae70d46dcef..cd1a2ae263da002db35be6bf5d937562f0dd8c52 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -662,6 +662,8 @@ public final class Bukkit {
         return server.getPlayer(name);
     }
 
+    public static Player getPlayerByIdentifier(@NotNull String id) { return server.getPlayerByIdentifier(id); }
+
     /**
      * Gets the player with the exact given name, case insensitive.
      *
diff --git a/src/test/java/org/bukkit/AnnotationTest.java b/src/test/java/org/bukkit/AnnotationTest.java
index a85b020763f55aa7200400f7abdc4b482514adf0..48b5fca79d96af1ba1c25705b0cebe5496b90a58 100644
--- a/src/test/java/org/bukkit/AnnotationTest.java
+++ b/src/test/java/org/bukkit/AnnotationTest.java
@@ -49,6 +49,8 @@ public class AnnotationTest {
         "org/bukkit/craftbukkit/CraftModule",
         "org/bukkit/craftbukkit/ServerModule",
         "org/bukkit/Server",
+        "org/bukkit/entity/Player",
+        "org/bukkit/Bukkit",
         "org/bukkit/plugin/java/ModuleJavaPlugin",
         "org/bukkit/plugin/Plugin",
         "org/bukkit/plugin/PluginModule",
