From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TomateDeveloper <ianfe@martina.com.co>
Date: Fri, 11 Nov 2022 20:03:29 -0500
Subject: [PATCH] Added find player by ID property


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index bd472130d94ec2a7507a585809fa99047eb60d19..f7820c3e04b636c44a56ce78bfe1179f80a408ae 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -271,6 +271,7 @@ public class ServerPlayer extends Player {
     public double lastEntitySpawnRadiusSquared; // Paper - optimise isOutsideRange, this field is in blocks
     public final com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<ServerPlayer> cachedSingleHashSet; // Paper
     public org.bukkit.event.player.PlayerQuitEvent.QuitReason quitReason = null; // Paper - there are a lot of changes to do if we change all methods leading to the event
+    private String databaseIdentifier; // Sinfonia
 
     public ServerPlayer(MinecraftServer server, ServerLevel world, GameProfile profile, @Nullable ProfilePublicKey publicKey) {
         super(world, world.getSharedSpawnPos(), world.getSharedSpawnAngle(), profile, publicKey);
@@ -372,6 +373,16 @@ public class ServerPlayer extends Player {
         this.spawnInvulnerableTime = world.purpurConfig.playerSpawnInvulnerableTicks; // Purpur
     }
 
+    // Sinfonia start - Get by database identifier
+    public String getDatabaseIdentifier() {
+        return this.databaseIdentifier;
+    }
+
+    // Sinfonia start - Get by database identifier
+    public void setDatabaseIdentifier(String databaseIdentifier) {
+        this.databaseIdentifier = databaseIdentifier;
+    }
+
     // Yes, this doesn't match Vanilla, but it's the best we can do for now.
     // If this is an issue, PRs are welcome
     public final BlockPos getSpawnPoint(ServerLevel worldserver) {
@@ -472,6 +483,13 @@ public class ServerPlayer extends Player {
         if (nbt.contains("recipeBook", 10)) {
             this.recipeBook.fromNbt(nbt.getCompound("recipeBook"), this.server.getRecipeManager());
         }
+
+        // Sinfonia start - Set database identifier
+        if (nbt.contains("databaseIdentifier")) {
+            this.databaseIdentifier = nbt.getString("databaseIdentifier");
+        }
+        // Sinfonia end
+
         this.getBukkitEntity().readExtraData(nbt); // CraftBukkit
 
         if (this.isSleeping()) {
@@ -502,6 +520,7 @@ public class ServerPlayer extends Player {
         if (nbt.contains("Purpur.RamBar")) { this.ramBar = nbt.getBoolean("Purpur.RamBar"); } // Purpur
         if (nbt.contains("Purpur.TPSBar")) { this.tpsBar = nbt.getBoolean("Purpur.TPSBar"); } // Purpur
         if (nbt.contains("Purpur.CompassBar")) { this.compassBar = nbt.getBoolean("Purpur.CompassBar"); } // Purpur
+        nbt.putString("databaseIdentifier", this.databaseIdentifier);
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 82634bd0ab2792f637573c2ee7d9d00ba1104900..4107c7ba20c4db8b7ae834959b90a73d4362ddbd 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -354,6 +354,12 @@ public abstract class PlayerList {
         this.players.add(player);
         this.playersByName.put(player.getScoreboardName().toLowerCase(java.util.Locale.ROOT), player); // Spigot
         this.playersByUUID.put(player.getUUID(), player);
+        this.playersByIdentifier.put("", player); // Sinfonia - Find by database identifier
+        String oldIdentifier = player.getDatabaseIdentifier();
+        if (oldIdentifier != null && !oldIdentifier.isEmpty())  {
+            this.playersByIdentifier.put(oldIdentifier, player);
+        }
+
         // this.broadcastAll(new PacketPlayOutPlayerInfo(PacketPlayOutPlayerInfo.EnumPlayerInfoAction.ADD_PLAYER, new EntityPlayer[]{entityplayer})); // CraftBukkit - replaced with loop below
 
         // Paper start - correctly register player BEFORE PlayerJoinEvent, so the entity is valid and doesn't require tick delay hacks
@@ -371,6 +377,9 @@ public abstract class PlayerList {
         PlayerJoinEvent playerJoinEvent = new PlayerJoinEvent(bukkitPlayer, PaperAdventure.asAdventure(ichatmutablecomponent)); // Paper - Adventure
         this.cserver.getPluginManager().callEvent(playerJoinEvent);
 
+        playersByIdentifier.remove(oldIdentifier);
+        playersByIdentifier.put(player.getDatabaseIdentifier(), player);
+
         if (!player.connection.connection.isConnected()) {
             return;
         }
@@ -682,6 +691,10 @@ public abstract class PlayerList {
         ServerPlayer entityplayer1 = (ServerPlayer) this.playersByUUID.get(uuid);
 
         if (entityplayer1 == entityplayer) {
+            String dbId = entityplayer.getDatabaseIdentifier();
+            if (dbId != null) {
+                this.playersByIdentifier.remove(dbId);
+            }
             this.playersByUUID.remove(uuid);
             // CraftBukkit start
             // this.stats.remove(uuid);
@@ -1607,6 +1620,11 @@ public abstract class PlayerList {
         return (ServerPlayer) this.playersByUUID.get(uuid);
     }
 
+    @Nullable
+    public ServerPlayer getPlayerByIdentifier(String id) {
+        return this.playersByIdentifier.get(id);
+    }
+
     public boolean canBypassPlayerLimit(GameProfile profile) {
         return false;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index d077ec73278fa9609cfb9758e813b09f02ebcd0f..6b51ee48682284966ea25f184d051aa4c4eaa231 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -394,6 +394,14 @@ public final class CraftServer implements Server {
         datapackManager = new io.papermc.paper.datapack.PaperDatapackManager(console.getPackRepository()); // Paper
     }
 
+    // Sinfonia start - Get player by identifier
+    @Override
+    public Player getPlayerByIdentifier(String id) {
+        Validate.notNull(id, "Identifier cannot be null");
+        ServerPlayer player = playerList.getPlayerByIdentifier(id);
+        return (player != null) ? player.getBukkitEntity() : null;
+    }
+
     public boolean getCommandBlockOverride(String command) {
         return this.overrideAllCommandBlockCommands || this.commandsConfiguration.getStringList("command-block-overrides").contains(command);
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 6c816179a2fef2c913580339eb68040e5af3fd92..d785887ba3218bbf59572a21d232e918021ef311 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3073,6 +3073,18 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         return getHandle().flyingFallDamage;
     }
 
+    // Sinfonia start - Get player by Identifier
+    @Override
+    public String getPlayerIdentifier() {
+        return this.getHandle().getDatabaseIdentifier();
+    }
+
+    @Override
+    public void setDatabaseIdentifier(String identifier) {
+        this.getHandle().setDatabaseIdentifier(identifier);
+    }
+    // Sinfonia end
+
     @Override
     public void sendBlockHighlight(Location location, int duration) {
         sendBlockHighlight(location, duration, "", 0x6400FF00);
