From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TomateDeveloper <ianfe@martina.com.co>
Date: Sat, 12 Nov 2022 13:39:54 -0500
Subject: [PATCH] Reordered database identified and assert not nulling


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 80286c84003437779a51d077f0f92783cd0b2abf..d88e29bea2294074e27ec55dad87beebed8c3428 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -273,7 +273,7 @@ public class ServerPlayer extends Player {
     private boolean ramBar = false; // Purpur
     private boolean tpsBar = false; // Purpur
     private boolean compassBar = false; // Purpur
-    private String databaseIdentifier; // Sinfonia
+    private String databaseIdentifier = ""; // Sinfonia
 
     public ServerPlayer(MinecraftServer server, ServerLevel world, GameProfile profile) {
         super(world, world.getSharedSpawnPos(), world.getSharedSpawnAngle(), profile);
@@ -497,12 +497,6 @@ public class ServerPlayer extends Player {
             this.recipeBook.fromNbt(nbt.getCompound("recipeBook"), this.server.getRecipeManager());
         }
 
-        // Sinfonia start - Set database identifier
-        if (nbt.contains("databaseIdentifier")) {
-            this.databaseIdentifier = nbt.getString("databaseIdentifier");
-        }
-        // Sinfonia end
-
         this.getBukkitEntity().readExtraData(nbt); // CraftBukkit
 
         if (this.isSleeping()) {
@@ -533,7 +527,10 @@ public class ServerPlayer extends Player {
         if (nbt.contains("Purpur.RamBar")) { this.ramBar = nbt.getBoolean("Purpur.RamBar"); } // Purpur
         if (nbt.contains("Purpur.TPSBar")) { this.tpsBar = nbt.getBoolean("Purpur.TPSBar"); } // Purpur
         if (nbt.contains("Purpur.CompassBar")) { this.compassBar = nbt.getBoolean("Purpur.CompassBar"); } // Purpur
+        if (nbt.contains("databaseIdentifier")) { this.databaseIdentifier = nbt.getString("databaseIdentifier"); } // Sinfonia
+
         nbt.putString("databaseIdentifier", this.databaseIdentifier);
+
     }
 
     @Override
