From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TomateDeveloper <ianfe@martina.com.co>
Date: Sun, 13 Nov 2022 18:42:20 -0500
Subject: [PATCH] New approach of set player identifier


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index d88e29bea2294074e27ec55dad87beebed8c3428..40946043f547311bfb130da8b378a27494a07ed2 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -384,6 +384,7 @@ public class ServerPlayer extends Player {
     // Sinfonia start - Get by database identifier
     public void setDatabaseIdentifier(String databaseIdentifier) {
         this.databaseIdentifier = databaseIdentifier;
+        this.server.getPlayerList().setPlayerWithIdentifier(this, databaseIdentifier);
     }
 
     // Yes, this doesn't match Vanilla, but it's the best we can do for now.
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index c2fe4723ec3098c8fb0d184c67bf08837afc80f5..fab17e1324c8e20141811c817fc38fd2d4885e5b 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -303,12 +303,6 @@ public abstract class PlayerList {
         this.playersByName.put(player.getScoreboardName().toLowerCase(java.util.Locale.ROOT), player); // Spigot
         this.playersByUUID.put(player.getUUID(), player);
 
-        this.playersByIdentifier.put("", player); // Sinfonia - Find by database identifier
-        String oldIdentifier = player.getDatabaseIdentifier();
-        if (oldIdentifier != null && !oldIdentifier.isEmpty())  {
-            this.playersByIdentifier.put(oldIdentifier, player);
-        }
-
         // this.broadcastAll(new PacketPlayOutPlayerInfo(PacketPlayOutPlayerInfo.EnumPlayerInfoAction.ADD_PLAYER, new EntityPlayer[]{entityplayer})); // CraftBukkit - replaced with loop below
 
         // Paper start - correctly register player BEFORE PlayerJoinEvent, so the entity is valid and doesn't require tick delay hacks
@@ -326,9 +320,6 @@ public abstract class PlayerList {
         PlayerJoinEvent playerJoinEvent = new PlayerJoinEvent(bukkitPlayer, PaperAdventure.asAdventure(ichatmutablecomponent)); // Paper - Adventure
         this.cserver.getPluginManager().callEvent(playerJoinEvent);
 
-        playersByIdentifier.remove(oldIdentifier);
-        playersByIdentifier.put(player.getDatabaseIdentifier(), player);
-
         if (!player.connection.connection.isConnected()) {
             return;
         }
@@ -644,10 +635,7 @@ public abstract class PlayerList {
         ServerPlayer entityplayer1 = (ServerPlayer) this.playersByUUID.get(uuid);
 
         if (entityplayer1 == entityplayer) {
-            String dbId = entityplayer.getDatabaseIdentifier();
-            if (dbId != null) {
-                this.playersByIdentifier.remove(dbId);
-            }
+            this.playersByIdentifier.remove(entityplayer.getDatabaseIdentifier()); // Sinfonia - Remove player on disconnect
             this.playersByUUID.remove(uuid);
             // CraftBukkit start
             // this.stats.remove(uuid);
@@ -1557,6 +1545,13 @@ public abstract class PlayerList {
         return this.playersByIdentifier.get(id);
     }
 
+    // Sinfonia start - Updated way of setting player identifier
+    public void setPlayerWithIdentifier(ServerPlayer player, String id) {
+        if (!id.equalsIgnoreCase("")) {
+            this.playersByIdentifier.put(id, player);
+        }
+    }
+
     public boolean canBypassPlayerLimit(GameProfile profile) {
         return false;
     }
