From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TomateDeveloper <ianfe@martina.com.co>
Date: Sun, 13 Nov 2022 18:42:20 -0500
Subject: [PATCH] New approach of set player identifier


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 1772ed63b42dafa3f9d9ab05f26981d2b039f457..4d05dec9b67bda0e4beead302de70e1002b50552 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -386,6 +386,7 @@ public class ServerPlayer extends Player {
     // Sinfonia start - Get by database identifier
     public void setDatabaseIdentifier(String databaseIdentifier) {
         this.databaseIdentifier = databaseIdentifier;
+        this.server.getPlayerList().setPlayerWithIdentifier(this, databaseIdentifier);
     }
 
     // Yes, this doesn't match Vanilla, but it's the best we can do for now.
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 13de9aa830c79000acca0c1bfa97a22fedfce71c..50f86bacace000077cbd7046a0969a1c925d0781 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -365,12 +365,6 @@ public abstract class PlayerList {
         this.playersByName.put(player.getScoreboardName().toLowerCase(java.util.Locale.ROOT), player); // Spigot
         this.playersByUUID.put(player.getUUID(), player);
 
-        this.playersByIdentifier.put("", player); // Sinfonia - Find by database identifier
-        String oldIdentifier = player.getDatabaseIdentifier();
-        if (oldIdentifier != null && !oldIdentifier.isEmpty())  {
-            this.playersByIdentifier.put(oldIdentifier, player);
-        }
-
         // this.broadcastAll(new PacketPlayOutPlayerInfo(PacketPlayOutPlayerInfo.EnumPlayerInfoAction.ADD_PLAYER, new EntityPlayer[]{entityplayer})); // CraftBukkit - replaced with loop below
 
         // Paper start - correctly register player BEFORE PlayerJoinEvent, so the entity is valid and doesn't require tick delay hacks
@@ -388,9 +382,6 @@ public abstract class PlayerList {
         PlayerJoinEvent playerJoinEvent = new PlayerJoinEvent(bukkitPlayer, PaperAdventure.asAdventure(ichatmutablecomponent)); // Paper - Adventure
         this.cserver.getPluginManager().callEvent(playerJoinEvent);
 
-        playersByIdentifier.remove(oldIdentifier);
-        playersByIdentifier.put(player.getDatabaseIdentifier(), player);
-
         if (!player.connection.connection.isConnected()) {
             return;
         }
@@ -701,10 +692,7 @@ public abstract class PlayerList {
         ServerPlayer entityplayer1 = (ServerPlayer) this.playersByUUID.get(uuid);
 
         if (entityplayer1 == entityplayer) {
-            String dbId = entityplayer.getDatabaseIdentifier();
-            if (dbId != null) {
-                this.playersByIdentifier.remove(dbId);
-            }
+            this.playersByIdentifier.remove(entityplayer.getDatabaseIdentifier()); // Sinfonia - Remove player on disconnect
             this.playersByUUID.remove(uuid);
             // CraftBukkit start
             // this.stats.remove(uuid);
@@ -1618,6 +1606,13 @@ public abstract class PlayerList {
         return this.playersByIdentifier.get(id);
     }
 
+    // Sinfonia start - Updated way of setting player identifier
+    public void setPlayerWithIdentifier(ServerPlayer player, String id) {
+        if (!id.equalsIgnoreCase("")) {
+            this.playersByIdentifier.put(id, player);
+        }
+    }
+
     public boolean canBypassPlayerLimit(GameProfile profile) {
         return false;
     }
